README.md
# linkScrape 
![Supported Python versions](https://img.shields.io/badge/python-2.7-blue.svg)

    linkScrape.py
    Description: Enumerates employee names from LinkedIn.com 
    Developers: Nick Sanzotta (@beamr), Jacob Robles (@shellfail)

**Considerations:**

    linkScrape is a pure Web Scraper, that does not utilize LinkedIn's API.
    linkScrape has limitations/bugs when scraping some character sets.
    Your LinkedIn.com account may be flagged or banned.
    Your LinkedIn.com account will need a minimum of 10 connections to to perform company based searches.
    Your LinkedIn.com account has a monthly commercial use limit.

**Installation:**

    git clone https://github.com/NickSanzotta/linkScrape.git
    cd linkScrape
    pip install -r requirements.txt
    python linkScrape.py --help


 **TIPS:**
 
**Use quotes for companies with white space**
 
 `python linkScrape.py -c'Example Company' -r1`
   
<br>

**LinkedIn Company ID search:**

Using a LinkedIn company ID is the most accurate way to search for a company. It's also the only way to search for universities.

`python linkScrape.py -c 100 -r1`

<br>

**Finding a LinkedIn company ID:**

In your browser perform a search for a company on LinkedIn.com. Once you find the company's profile page click the link: 

"<i>See all (Numeral) employees on LinkedIn</i>" 

Next inspect the URL for the LinkedIn company ID, below is an example URL with the company ID highlighted in bold.
<br>

ht<span>tps://</span><span>ww</span>w.linkedin.com/search/results/people/?facetCurrentCompany=%5B%22<b>39624</b>%22%2C%22118552%22%5D&lipi=urn%3Ali%3Apage%3Ad_flagship3_company%3BdmKCXJhuRE2mHw1V0%2BqXhw%3D%3D
  
<br>

**Similar named companies:**

Companies that share similar names, will produce multiple results and require the user to choose one, as shown in the example below.

```
python linkScrape.py -c'Example Company' -r1
  
1: Example Company
2: Example Inc.
3: Example Advertising
            
Please Select a Company: 1
ENTERED: "Example Company"

Company Name: Example Company
Industry / Employees: Printing; 5001-10,000 employees
LinkedIn CompanyID: 100
URL: https://www.linkedin.com/company/100
```
<br>
**Default Values:**

    If a parameter is not defined it's default value will be choosen.
    Default values listed below.
  
    Mangle Option = 7  ex: FLast
    Page Results = 1-3
    Time out value = 3
    
**Usage (CLI):**

    Usage: python linkScrape.py <OPTIONS>
    Example[1]: python linkScrape.py -e LinkedInUser@email.com -c 'Example Company' -r 10 -t 3 -m 7 -d example.com
    Example[2]: python linkScrape.py -e LinkedInUser@email.com -c 'Example Company' -r 5-10 -t 3 -m 7 -d example.com
    Example[3]: python linkScrape.py -e LinkedInUser@email.com -c 100 -r 3 -t 3 -m 7 -d example.com
    Example[4]: python linkScrape.py -m 7 -i ~/Company/names.txt\n"
    Formatted output saved to: linkedIn/linkScrape-data/Company-mangle[x]_time.txt
    
    Login options:
    -e <email> Your LinkedIn.com Email Address.
    -p <pass>  Your LinkedIn.com Password. (If -p parameter is not defined, you'll be prompt to enter a password)
    
    Search options:
    -c <company> Search company name or company ID.
    -r <results> Searches X number of LinkedIn.com pages, or a range of pages (Default is 1-3).
    -t <secs>    Sets timeout value. (Default is 3.)
 
**Usage (Wizard):**

      ENTERED: "Example Company"


       Mangle options:

             -m <mangle>    
                                       1)FirstLast        
                                       2)LastFirst        
                                       3)First.Last       
                                       4)Last.First       
                                       5)First_Last       
                                       6)Last_First       
                                       7)FLast            
                                       8)LFirst           
                                       9)FirstL           
                                      10)F.Last           
                                      11)L.Firstname      
                                      12)FirLa            
                                      13)Lastfir
                                      14)FirstLastnam
                                      15)LastF
                                      16)LasFi
                                      99)All              Mangle using all types

      Enter name Managle choice[ex:7]: 
      ENTERED: "7"

      [*]TIP: This value will determine how many page results will be returned.
      Enter number of pages results[ex:3] or a range of pages [ex:1-3]: 
      ENTERED: "1-3"

      [*]TIP: This value will determine how long of a delay(in seconds) each page will be scraped.
      Enter timeout value[ex:3]: 
      ENTERED: "3"

      [*]TIP: This value will be added to the end of each mangled result[ex:jdoe@example.com].
      Enter Domain suffix[ex:example.com]: example.com
      ENTERED: "example.com"


**Mangle Options:** 
    
    -m <mangle>
        1)FirstLast        
        2)LastFirst        
        3)First.Last       
        4)Last.First       
        5)First_Last       
        6)Last_First       
        7)FLast            
        8)LFirst           
        9)FirstL           
        10)F.Last          
        11)L.Firstname     
        12)FirLa           
        13)Lastfir
        14)FirstLastnam             
        15)LastF
        16)LasFi
        99)All              Mangle using all types

  
    -d <domain> Append @example.com to enumerated user list."
    -i <input>  Use local file instead of LinkedIn.com to perform name Mangle."
    
    -h <help>  Prints this help menu.


changelog
06/24/2018
------------
- Version 2.1 Master release
	- Public release
	- Fixed issue #30, Names with apostrophes or certs/titles in names fields aren't correctly mangled.
	- Partially Fixed issue #45, Only write and display unique names to logs.

06/21/2018
------------
- Version 2.0 Master release
	- Added LinkedIn Company ID search.
	- Added verbosity to STDOUT
	- Refactored code to include LinkScraper class.
	- Fixed issue #46, added CTRL C for gracefully exit /w mangle completion.


linkMangle.py
#!/usr/bin/env python2
import os.path
import time

# Mangle timestamp.
Mangle_timestamp = time.strftime("%m-%d-%Y_%H:%M")

def write(companyName, newname, first_name, last_name, filename):
    with open(filename, 'a') as f:
        f.write(newname+"\n")

    directory_path = '%s%s' % ('logs/', companyName)
    if not os.path.exists(directory_path):
        os.mkdir(directory_path)
    # Write to csv file, supports msfpro phishing campaigns format.
    msf_filename = '%s%s.csv' % ('msfpro_', Mangle_timestamp)
    msfpath = os.path.join(directory_path, msf_filename)
    
    if not os.path.isfile(msfpath):
        with open(msfpath, 'a') as f1:
            f1.write('email_address,first_name,last_name\n')
    
    with open(msfpath, 'a') as f2:
        f2.write(newname+','+first_name.title()+','+last_name.title()+'\n')

def mangleOne(names, companyName, domain, filename):
    newnames = []
    for name in names:
        first_name = name[0]
        last_name = name[1]
        newname=first_name + last_name
        if domain != '':
            newname = newname+"@"+domain
            write(companyName, newname, first_name, last_name, filename)
        else:
            write(companyName, newname, first_name, last_name, filename)
        newnames.append(newname)
    return newnames

def mangleTwo(names, companyName, domain, filename):
    newnames = []
    for name in names:
        first_name = name[0]
        last_name = name[1]
        newname = last_name + first_name
        if domain != '':
            newname = newname+"@"+domain
            write(companyName, newname, first_name, last_name, filename)
        else:
            write(companyName, newname, first_name, last_name, filename)
        newnames.append(newname)
    return newnames

def mangleThree(names, companyName, domain, filename):
    newnames = []
    for name in names:
        first_name = name[0]
        last_name = name[1]
        newname = first_name + "." + last_name
        if domain != '':
            newname = newname+"@"+domain
            write(companyName, newname, first_name, last_name, filename)
        else:
            write(companyName, newname, first_name, last_name, filename)
        newnames.append(newname)
    return newnames


def mangleFour(names, companyName, domain, filename):
    newnames = []
    for name in names:
        first_name = name[0]
        last_name = name[1]
        newname = last_name + "." + first_name
        if domain != '':
            newname = newname+"@"+domain
            write(companyName, newname, first_name, last_name, filename)
        else:
            write(companyName, newname, first_name, last_name, filename)
        newnames.append(newname)
    return newnames

def mangleFive(names, companyName, domain, filename):
    newnames = []
    for name in names:
        first_name = name[0]
        last_name = name[1]
        newname = first_name + "_" + last_name
        if domain != '':
            newname = newname+"@"+domain
            write(companyName, newname, first_name, last_name, filename)
        else:
            write(companyName, newname, first_name, last_name, filename)
        newnames.append(newname)
    return newnames

def mangleSix(names, companyName, domain, filename):
    newnames = []
    for name in names:
        first_name = name[0]
        last_name = name[1]
        newname = last_name + "_" + first_name
        if domain != '':
            newname = newname+"@"+domain
            write(companyName, newname, first_name, last_name, filename)
        else:
            write(companyName, newname, first_name, last_name, filename)
        newnames.append(newname)
    return newnames


def mangleSeven(names, companyName, domain, filename):
    newnames = []
    for name in names:
        first_name = name[0]
        last_name = name[1]
        newname = first_name[0] + last_name
        if domain != '':
            newname = newname+"@"+domain
            write(companyName, newname, first_name, last_name, filename)
        else:
            write(companyName, newname, first_name, last_name, filename)
        newnames.append(newname)
    return newnames

def mangleEight(names, companyName, domain, filename):
    newnames = []
    for name in names:
        first_name = name[0]
        last_name = name[1]
        newname = last_name[0] + first_name
        if domain != '':
            newname = newname+"@"+domain
            write(companyName, newname, first_name, last_name, filename)
        else:
            write(companyName, newname, first_name, last_name, filename)
        newnames.append(newname)
    return newnames

def mangleNine(names, companyName, domain, filename):
    newnames = []
    for name in names:
        first_name = name[0]
        last_name = name[1]
        newname = first_name + last_name[0]
        if domain != '':
            newname = newname+"@"+domain
            write(companyName, newname, first_name, last_name, filename)
        else:
            write(companyName, newname, first_name, last_name, filename)
        newnames.append(newname)
    return newnames


def mangleTen(names, companyName, domain, filename):
    newnames = []
    for name in names:
        first_name = name[0]
        last_name = name[1]
        newname = first_name[0] + "." + last_name
        if domain != '':
            newname = newname+"@"+domain
            write(companyName, newname, first_name, last_name, filename)
        else:
            write(companyName, newname, first_name, last_name, filename)
        newnames.append(newname)
    return newnames

def mangleEleven(names, companyName, domain, filename):
    newnames = []
    for name in names:
        first_name = name[0]
        last_name = name[1]
        newname = last_name[0] + "." + first_name
        if domain != '':
            newname = newname+"@"+domain
            write(companyName, newname, first_name, last_name, filename)
        else:
            write(companyName, newname, first_name, last_name, filename)
        newnames.append(newname)
    return newnames

def mangleTwelve(names, companyName, domain, filename):
    newnames = []
    for name in names:
        first_name = name[0]
        last_name = name[1]
        newname = last_name[0:3] + first_name[0:2]
        if domain != '':
            newname = newname+"@"+domain
            write(companyName, newname, first_name, last_name, filename)
        else:
            write(companyName, newname, first_name, last_name, filename)
        newnames.append(newname)
    return newnames


def mangleThirteen(names, companyName, domain, filename):
    newnames = []
    for name in names:
        first_name = name[0]
        last_name = name[1]
        newname = last_name[0:4] + first_name[0:3]
        if domain != '':
            newname = newname+"@"+domain
            write(companyName, newname, first_name, last_name, filename)
        else:
            write(companyName, newname, first_name, last_name, filename)
        newnames.append(newname)
    return newnames

def mangleFourteen(names, companyName, domain, filename):
    newnames = []
    for name in names:
        first_name = name[0]
        last_name = name[1]
        newname = first_name[0] + last_name[0:7]
        if domain != '':
            newname = newname+"@"+domain
            write(companyName, newname, first_name, last_name, filename)
        else:
            write(companyName, newname, first_name, last_name, filename)
        newnames.append(newname)
    return newnames

def mangleFifteen(names, companyName, domain, filename):
    newnames = []
    for name in names:
        first_name = name[0]
        last_name = name[1]
        newname = last_name + first_name[0]
        if domain != '':
            newname = newname+"@"+domain
            write(companyName, newname, first_name, last_name, filename)
        else:
            write(companyName, newname, first_name, last_name, filename)
        newnames.append(newname)
    return newnames

def mangleSixteen(names, companyName, domain, filename):
    newnames = []
    for name in names:
        first_name = name[0]
        last_name = name[1]
        newname = last_name[0:3] + first_name[0:2]
        if domain != '':
            newname = newname+"@"+domain
            write(companyName, newname, first_name, last_name, filename)
        else:
            write(companyName, newname, first_name, last_name, filename)
        newnames.append(newname)
    return newnames

def mangleAll(names, companyName, domain, filename):
    newnames = []
    newname = mangleOne(names, companyName, domain, filename)
    print(newname)
    newname = mangleTwo(names, companyName, domain, filename)
    print(newname)
    newname = mangleThree(names, companyName, domain, filename)
    print(newname)
    newname = mangleFour(names, companyName, domain, filename)
    print(newname)
    newname = mangleFive(names, companyName, domain, filename)
    print(newname)
    newname = mangleSix(names, companyName, domain, filename)
    print(newname)
    newname = mangleSeven(names, companyName, domain, filename)
    print(newname)
    newname = mangleEight(names, companyName, domain, filename)
    print(newname)
    newname = mangleNine(names, companyName, domain, filename)
    print(newname)
    newname = mangleTen(names, companyName, domain, filename)
    print(newname)
    newname = mangleEleven(names, companyName, domain, filename)
    print(newname)
    newname = mangleTwelve(names, companyName, domain, filename)
    print(newname)
    newname = mangleThirteen(names, companyName, domain, filename)
    print(newname)
    newname = mangleFourteen(names, companyName, domain, filename)
    print(newname)
    newname = mangleFifteen(names, companyName, domain, filename)
    print(newname)
    newname = mangleSixteen(names, companyName, domain, filename)
    print(newname)

if __name__ == '__main__':
    pass

linkArguments.py
#!/usr/bin/env python2
"""linkScrape Arguments."""
import argparse
import getpass
import sys
import time
from argparse import RawTextHelpFormatter

Version = 'v2.1'

timestr = time.strftime("%Y%m%d-%H%M")


class Colors:
    """Terminal color options."""

    white = "\033[1;37m"
    normal = "\033[0;00m"
    red = "\033[1;31m"
    blue = "\033[1;34m"
    green = "\033[1;32m"
    lightblue = "\033[0;34m"

banner = (Colors.lightblue + r"""
 ___                __      ____
/\_ \    __        /\ \    /\  _`\
\//\ \  /\_\    ___\ \ \/'\\ \,\L\_\    ___   _ __    __     _____      __
  \ \ \ \/\ \ /' _ `\ \ , < \/_\__ \   /'___\/\`'__\/'__`\  /\ '__`\  /'__`\
   \_\ \_\ \ \/\ \/\ \ \ \\`\ /\ \L\ \/\ \__/\ \ \//\ \L\.\_\ \ \L\ \/\  __/
   /\____\\ \_\ \_\ \_\ \_\ \_\ `\____\ \____\\ \_\\ \__/.\_\\ \ ,__/\ \____\
   \/____/ \/_/\/_/\/_/\/_/\/_/\/_____/\/____/ \/_/ \/__/\/_/ \ \ \/  \/____/
                                                               \ \_\
                                                                \/_/
""" + 
'\nlinkScrape.py ' + Version + Colors.normal +
'\nDescription: Enumerates employee names from LinkedIn.com ' +
'\nDevelopers: Nick Sanzotta & Jacob Robles.\n' + '*' * 80 + '\n')


def valid_range(pages):
    """..."""
    pages = pages.strip()
    if pages.isdigit():
        # Use this as last page
        return
    elif '-' in pages:
        # May have a range
        pages = pages.split('-')
        if len(pages) != 2:
            print 'Invalid Range'
            sys.exit(1)
        else:
            first, last = pages[0].strip(), pages[1].strip()
            if first.isdigit() and last.isdigit() and int(first) >= 1 and int(first) < int(last):
                # print int(first), int(last), 'Valid'
                pass
            else:
                print 'Invalid Range'
                sys.exit(1)
    else:
        # Invalid range
        print 'Invalid Range'
        sys.exit(1)


def parse_args():
    """Argparse arguments."""
    parser = argparse.ArgumentParser(formatter_class=RawTextHelpFormatter, description=banner +
    "Usage: python linkScrape.py <OPTIONS> \n" +
    "Example 1: python linkScrape.py -e LinkedInUser@email.com -c 'example company' -r 1-3 -t 15 -m 10 -d example.com\n" +
    "Example 2: python linkScrape.py -e LinkedInUser@email.com -c 'example company' -r 3 -t 15 -m 10 -d example.com\n" +
    "Example 3: python linkScrape.py -m 7 -i ~/Company/names.txt\n")

    auth_group = parser.add_argument_group(Colors.lightblue + 'Login options' + Colors.normal)
    auth_group.add_argument('-e', '--email', help="Your LinkedIn.com Email Address")
    auth_group.add_argument('-p', '--password', help="Your LinkedIn.com Password.\n" + "If -p is not defined, you'll be prompted")

    search_group = parser.add_argument_group(Colors.lightblue + 'Search options' + Colors.normal)
    search_group.add_argument('-c', '--company', help="Search company name or company ID.")
    search_group.add_argument('-u', '--university', action='count', help="University flag if searching for university. (Default is False)")
    search_group.add_argument('-r', '--range', default='1-3', help="The number/range of page searches.\n" + "(Default is 1-3)")
    search_group.add_argument('-t', '--timeout', type=int, default=3, help="Timeout value. (Default is 3)")
    mangle_help = (
        "1)FirstLast\n"
        "2)LastFirst\n"
        "3)First.Last\n"
        "4)Last.First\n"
        "5)First_Last\n"
        "6)Last_First\n"
        "7)FLast\n"
        "8)LFirst\n"
        "9)FirstL\n"
        "10)F.Last\n"
        "11)L.Firstname\n"
        "12)FirLa\n"
        "13)LastFir\n"
        "14)FLastnam\n"
        "15)LastF\n"
        "16)LasFi\n"
        "99)All\tMangle using all types")

    mangle_group = parser.add_argument_group(Colors.lightblue + 'Mangle options' + Colors.normal)
    mangle_group.add_argument('-m', '--mangle', type=int, choices=list(range(1, 17)) + [99], default=7, help=mangle_help)
    mangle_group.add_argument('-d', '--domain', help='Append @domain.com to enumerater user list')
    mangle_group.add_argument('-i', '--input', help='Use local file to perform name Mangle.')

    args = parser.parse_args()

    # Only ask for user/pass when doing LinkedIn search
    if not args.input:
        if not args.email:
            args.email = raw_input('Enter LinkedIn Email account[user@email.com]:')
            print('ENTERED: {0}\n'.format(args.email))
        if not args.password:
            args.password = getpass.getpass('Enter LinkedIn Password: ')
    while not args.company:
        args.company = raw_input('Enter Company[ex:Example Company]: ').strip()
        print('ENTERED: {0}\n'.format(args.company))
    if not args.domain:
        args.domain = raw_input('Enter Domain suffix[ex:example.com]: ')
        print('ENTERED: {0}\n'.format(args.domain))
    valid_range(args.range)

    return args

if __name__ == "__main__":
    args = parse_args()
    print args
